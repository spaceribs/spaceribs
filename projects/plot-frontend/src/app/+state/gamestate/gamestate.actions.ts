import { createAction, props } from '@ngrx/store';
import { GameState } from './gamestate.models';

/**
 * Action to initialize gamestate.
 */
export const init = createAction('[GameState] Init');

/**
 * Action taken when successfully loading gamestate.
 */
export const loadGameStateSuccess = createAction(
  '[GameState/App] Load GameState Success',
  props<{ gamestate: GameState }>(),
);

/**
 * Action taken when failing to load gamestate.
 */
export const loadGameStateFailure = createAction(
  '[GameState/App] Load GameState Failure',
  props<{ error: unknown }>(),
);

/**
 * Action taken when refreshing the gamestate.
 */
export const refresh = createAction('[GameState/App] Refresh');
