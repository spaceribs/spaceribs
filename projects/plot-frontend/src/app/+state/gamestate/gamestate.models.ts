/**
 * Interface for the 'GameState' response
 */
export interface GameState {
  system: {
    time: string;
  };
  round: {
    end_time: string;
  };
  user: {
    units: number;
    username: string;
    avatar: string;
    nationId: number;
  } | null;
}
