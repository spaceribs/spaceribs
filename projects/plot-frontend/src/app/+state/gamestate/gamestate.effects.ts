import { HttpClient } from '@angular/common/http';
import { ErrorHandler, Injectable } from '@angular/core';
import { createEffect, Actions, ofType } from '@ngrx/effects';
import { fetch } from '@nrwl/angular';
import { map } from 'rxjs/operators';

import * as GameStateActions from './gamestate.actions';
import { GameState } from './gamestate.models';

/**
 * Observable effects associated with the gamestate feature.
 */
@Injectable()
export class GameStateEffects {
  init$ = createEffect(() =>
    this.actions$.pipe(
      ofType(GameStateActions.init, GameStateActions.refresh),
      fetch({
        run: () =>
          this.http
            .get<GameState>('/gamestate/')
            .pipe(
              map((res) =>
                GameStateActions.loadGameStateSuccess({ gamestate: res }),
              ),
            ),
        onError: (_, error: unknown) => {
          this.errorHandler.handleError(error);
          return GameStateActions.loadGameStateFailure({ error });
        },
      }),
    ),
  );

  constructor(
    private readonly actions$: Actions,
    private readonly errorHandler: ErrorHandler,
    private readonly http: HttpClient,
  ) {}
}
