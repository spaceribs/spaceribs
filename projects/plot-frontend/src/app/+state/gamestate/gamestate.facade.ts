import { Injectable } from '@angular/core';
import { Store } from '@ngrx/store';

import * as GameStateActions from './gamestate.actions';
import * as GameStateSelectors from './gamestate.selectors';

/**
 * Facade for all calls to and from the gamestate endpoints.
 */
@Injectable()
export class GameStateFacade {
  loaded$ = this.store$.select(GameStateSelectors.selectGameStateLoaded);
  gamestate$ = this.store$.select(GameStateSelectors.selectGameState);

  constructor(private readonly store$: Store) {}

  public init() {
    this.store$.dispatch(GameStateActions.init());
  }

  public refresh() {
    this.store$.dispatch(GameStateActions.refresh());
  }
}
