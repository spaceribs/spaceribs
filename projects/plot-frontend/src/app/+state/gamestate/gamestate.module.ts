import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { StoreModule } from '@ngrx/store';
import { EffectsModule } from '@ngrx/effects';
import * as fromGameState from './gamestate.reducer';
import { GameStateEffects } from './/gamestate.effects';
import { GameStateFacade } from './/gamestate.facade';

@NgModule({
  declarations: [],
  imports: [
    CommonModule,
    StoreModule.forFeature(
      fromGameState.GAMESTATE_FEATURE_KEY,
      fromGameState.reducer,
    ),
    EffectsModule.forFeature([GameStateEffects]),
  ],
  providers: [GameStateFacade],
})
export class GameStateModule {}
