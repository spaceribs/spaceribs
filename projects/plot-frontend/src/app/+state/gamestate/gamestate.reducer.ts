import { createReducer, on, Action } from '@ngrx/store';

import * as GameStateActions from './gamestate.actions';
import { GameState } from './gamestate.models';

/**
 * The feature key for gamestate.
 */
export const GAMESTATE_FEATURE_KEY = 'gamestate';

/**
 * State of the gamestate feature.
 */
export interface State {
  gamestate: GameState | null;
  loaded: boolean;
  error?: unknown | null;
}

/**
 * State of gamestate combined with other global states.
 */
export interface GameStatePartialState {
  readonly [GAMESTATE_FEATURE_KEY]: State;
}

/**
 * Initial state of gamestate.
 */
export const initialState: State = {
  gamestate: null,
  loaded: false,
  error: null,
};

const gamestateReducer = createReducer(
  initialState,
  on(
    GameStateActions.init,
    (state): State => ({ ...state, loaded: false, error: null }),
  ),
  on(
    GameStateActions.refresh,
    (state): State => ({ ...state, loaded: false, error: null }),
  ),
  on(
    GameStateActions.loadGameStateSuccess,
    (state, { gamestate }): State => ({
      ...state,
      gamestate,
      loaded: true,
    }),
  ),
  on(
    GameStateActions.loadGameStateFailure,
    (state, { error }): State => ({ ...state, error, loaded: true }),
  ),
);

/**
 * Reduce state by action.
 *
 * @param {(State | undefined)} state State to reduce
 * @param {Action} action Action to reduce by
 * @returns {State} State reduced by action
 */
export const reducer = (state: State | undefined, action: Action): State =>
  gamestateReducer(state, action);
