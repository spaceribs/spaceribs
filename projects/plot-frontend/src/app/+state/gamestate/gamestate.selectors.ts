import { createFeatureSelector, createSelector } from '@ngrx/store';
import { GAMESTATE_FEATURE_KEY, State } from './gamestate.reducer';

/**
 * Reduce the global state to gamestate.
 */
export const selectGameStateState = createFeatureSelector<State>(
  GAMESTATE_FEATURE_KEY,
);

/**
 * Reduce the gamestate state to loaded state.
 */
export const selectGameStateLoaded = createSelector(
  selectGameStateState,
  (state: State) => state.loaded,
);

/**
 * Reduce the gamestate state to error state.
 */
export const selectGameStateError = createSelector(
  selectGameStateState,
  (state: State) => state.error,
);

/**
 * Reduce the gamestate state to just the gamestate.
 */
export const selectGameState = createSelector(
  selectGameStateState,
  (state: State) => state.gamestate,
);
