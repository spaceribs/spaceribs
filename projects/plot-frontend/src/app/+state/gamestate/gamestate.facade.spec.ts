import * as GameStateActions from './gamestate.actions';

import {
  HttpClientTestingModule,
  HttpTestingController,
} from '@angular/common/http/testing';
import { GAMESTATE_FEATURE_KEY, reducer } from './gamestate.reducer';
import { Store, StoreModule } from '@ngrx/store';

import { EffectsModule } from '@ngrx/effects';
import { NgModule } from '@angular/core';
import { GameState } from './gamestate.models';
import { GameStateEffects } from './gamestate.effects';
import { GameStateFacade } from './gamestate.facade';
import { TestBed } from '@angular/core/testing';
import { firstValueFrom } from 'rxjs';

describe('GameStateFacade', () => {
  let facade: GameStateFacade;
  let store$: Store;
  let http: HttpTestingController;

  const createGameState = (): GameState => ({
    system: {
      time: '2024-01-13T20:30:50.167Z',
    },
    round: {
      end_time: '2024-01-13T23:30:50.167Z',
    },
    user: null,
  });

  describe('used in NgModule', () => {
    beforeEach(() => {
      @NgModule({
        imports: [
          StoreModule.forFeature(GAMESTATE_FEATURE_KEY, reducer),
          EffectsModule.forFeature([GameStateEffects]),
          HttpClientTestingModule,
        ],
        providers: [GameStateFacade],
      })
      class CustomFeatureModule {}

      @NgModule({
        imports: [
          StoreModule.forRoot({}),
          EffectsModule.forRoot([]),
          CustomFeatureModule,
        ],
      })
      class RootModule {}
      TestBed.configureTestingModule({ imports: [RootModule] });

      store$ = TestBed.inject<Store<object>>(Store);
      facade = TestBed.inject<GameStateFacade>(GameStateFacade);
      http = TestBed.inject(HttpTestingController);
    });

    /**
     * The initially generated facade::loadAll() returns empty array
     */
    it('loadAll() should return gamestate with loaded == true', async () => {
      let gamestate = await firstValueFrom(facade.gamestate$);
      let isLoaded = await firstValueFrom(facade.loaded$);

      expect(gamestate).toMatchSnapshot();
      expect(isLoaded).toBe(false);

      facade.init();

      http.expectOne('/gamestate/').flush([]);

      gamestate = await firstValueFrom(facade.gamestate$);
      isLoaded = await firstValueFrom(facade.loaded$);

      expect(gamestate).toHaveLength(0);
      expect(isLoaded).toBe(true);
    });

    /**
     * Use `loadGameStateSuccess` to manually update list
     */
    it('should start allGameState$ return the gamestate; and loaded flag == true', async () => {
      let gamestate = await firstValueFrom(facade.gamestate$);
      let isLoaded = await firstValueFrom(facade.loaded$);

      expect(gamestate).toBeNull();
      expect(isLoaded).toBe(false);

      store$.dispatch(
        GameStateActions.loadGameStateSuccess({
          gamestate: createGameState(),
        }),
      );

      gamestate = await firstValueFrom(facade.gamestate$);
      isLoaded = await firstValueFrom(facade.loaded$);

      expect(gamestate).toMatchInlineSnapshot();
      expect(isLoaded).toBe(true);
    });

    it('should call refresh', () => {
      const actionSpy = jest.spyOn(GameStateActions, 'refresh');
      const spy = jest.spyOn(store$, 'dispatch');
      facade.refresh();

      expect(spy).toHaveBeenCalled();
      expect(actionSpy).toHaveBeenCalled();
    });
  });
});
