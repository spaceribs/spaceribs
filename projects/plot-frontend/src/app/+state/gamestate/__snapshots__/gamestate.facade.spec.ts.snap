// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameStateFacade used in NgModule loadAll() should return empty list with loaded == true 1`] = `null`;

exports[`GameStateFacade used in NgModule loadAll() should return gamestate with loaded == true 1`] = `null`;
